### PR TITLE
[MRG+1] Add short-circuit return to matplotlib.artist.setp if input is length 0

### LINKED ISSUE #7784: Simple IndexError in artist.setp if empty list is passed
An IndexError is raised in matplotlib.artist.setp if argument is an empty list,

```
import matplotlib
matplotlib.artist.setp([])
```

Arose in Seaborn jointplot example in their docs where minor_ticks was empty.

Ubuntu 16.04, Python 3.4.3 and Matplotlib 2.0.0rc2+2754.g8c5b548.

### PR DESCRIPTION
**Pull request by samsontmr**
Fixes #7784 

Before this patch, an IndexError is raised in matplotlib.artist.setp if argument is an empty list,

```
import matplotlib
matplotlib.artist.setp([])
```

### CODE REVIEW

File: lib/matplotlib/artist.py [hunk 1]
```diff
@@ -1421,6 +1421,10 @@ def setp(obj, *args, **kwargs):
       >>> setp(lines, 'linewidth', 2, 'color', 'r')  # MATLAB style
       >>> setp(lines, linewidth=2, color='r')        # python style
     """
+    
+    if ((isinstance(obj, np.ndarray) and not obj.size) or
```
Comment (anntzer) [inline lib/matplotlib/artist.py]: Just check `if not objs:` instead (`objs` is a *list* defined immediately below).

**Comment by samsontmr** [comment]
Wouldn't `if not objs` return false if `objs` is a list of empty lists?

File: lib/matplotlib/artist.py (hunk 1 above)
Comment (samsontmr) [inline lib/matplotlib/artist.py]: Wouldn't `if not objs` return false if `objs` is a list of empty lists? Or do you mean to check for `if not obj`?

Comment (tacaswell) [inline lib/matplotlib/artist.py]: Agree using `if not objs:` would be better.  I would also move this down to L1453 (right above the line with `objs[0]`) after we have normalized the input to a list.

Comment (samsontmr) [inline lib/matplotlib/artist.py]: Oh I see! The if else statements guarantee a list, right?

Comment (tacaswell) [inline lib/matplotlib/artist.py]: Yes.

**Comment by samsontmr** [comment]
Commits squashed

**Comment by tacaswell** [comment]
Looks good!  Could you also add a test to `test_artist.py` that exercises this code path?

**Comment by phobson** [comment]
Apologies if I'm missing something obvious here, but couldn't this be at the very top of the function? e.g.,

```python

    if not objs:
        return
    elif not cbook.iterable(obj):
        objs = [obj]
    else:
        objs = list(cbook.flatten(obj))
```

**Comment by tacaswell** [comment]
@phobson users will find pathological cases :wink: 

```python
In [62]: bool([[]])
Out[62]: 
True
```

**Comment by phobson** [comment]
@tacaswell wow. great catch with that.

**Comment by samsontmr** [comment]
@phobson `objs` doesn't exist at the top either, only `obj` ;)

**Comment by samsontmr** [comment]
@tacaswell Tests added

**Comment by NelleV** [comment]
I've edited your comment to explain what it fixes. It saves us reviewers a lot of time not to have to identify what the original problem was.

**Comment by tacaswell** [comment]
This should be backported after 2.0

@samsontmr Thank!

**Comment by QuLogic** [comment]
This doesn't backport that easily, or at least, I'm not familiar enough with that section of the code to do it.

**Comment by tacaswell** [comment]
@QuLogic Lets drop the backport on this.

**Comment by QuLogic** [comment]
Works for me.

**Pull request merged**